src/software/thunderscope/gl/layers/gl_referee_info_layer.py

```diff
@@ -0,0 +1,249 @@
+        self.referee_vis_buffer = ThreadSafeBuffer(buffer_size, Referee)
+        self.ball_placement_vis_buffer = ThreadSafeBuffer(
+            buffer_size, BallPlacementVisualization
+        )
+        self.world_buffer = ThreadSafeBuffer(buffer_size, World)
+
+        self.cached_world = None
+        self.cached_referee_info = None
+
+        self.referee_text_graphics = ObservableList(self._graphics_changed)
+
+        self.placement_tolerance_graphic = GLCircle(
+            parent_item=self,
+            radius=BALL_PLACEMENT_TOLERANCE_RADIUS_METERS,
+            outline_color=self.BALL_PLACEMENT_TOLERANCE_VISUALIZATION_COLOR_OFF,
+        )
+
+        self.placement_target_graphic = GLCircle(
+            parent_item=self,
+            radius=BALL_PLACEMENT_TOLERANCE_RADIUS_METERS,
+            outline_color=self.BALL_PLACEMENT_TARGET_VISUALIZATION_COLOR,
+        )
+
+        self.robot_avoid_circle_graphic = GLCircle(
+            parent_item=self,
+            radius=BALL_PLACEMENT_ROBOT_AVOID_RADIUS_METERS,
+            outline_color=self.BALL_PLACEMENT_ROBOT_AVOID_AREA_VISUALIZATION_COLOR,
+        )
+
+        # initialize the two text items to display
+        self.gamestate_type_text: Optional[GLLabel] = None
+        self.command_type_text: Optional[GLLabel] = None
+
+        self.ball_placement_countdown_graphic: Optional[GLTextItem] = None
+
+        self.ball_placement_point = None
+        self.ball_placement_point_hidden = False
+        self.shrink_target = True
+        self.placement_start_time = 0
+
+    def __update_ball_placement(self) -> None:
+        """Update ball placement visuals"""
+        ball_placement_vis_proto = self.ball_placement_vis_buffer.get(
+            block=False, return_cached=False
+        )
+
+        if not self.ball_placement_countdown_graphic:
+            self.ball_placement_countdown_graphic = GLTextItem(
+                parentItem=self,
+                font=QtGui.QFont("Roboto", 7, weight=700),
+                color=self.COUNT_DOWN_TEXT_COLOR,
+            )
+
+        if not ball_placement_vis_proto and not self.ball_placement_point_hidden:
+            self.placement_tolerance_graphic.hide()
+            self.placement_target_graphic.hide()
+            self.robot_avoid_circle_graphic.hide()
+            self.ball_placement_countdown_graphic.hide()
+
+            self.ball_placement_point_hidden = True
+            self.shrink_target = True
+            return
+
+        if ball_placement_vis_proto:
+            # move the ball placement graphics to the new point
+            new_placement_point = ball_placement_vis_proto.ball_placement_point
+
+            # update the color of the target circle according to the position of the ball.
+            # if the ball lies inside the tolerance circle, the circle will be green, otherwise red.
+            ball_state = self.cached_world.ball.current_state
+            if not self.ball_placement_point_hidden:
```

Comment:
nit: would be better to pull out each branch of this if statement into 2 helper methods

so lines 131-146 in one, and lines 148-186 in one